### PR TITLE
Change default WKT output to honor a space between geometry type and Z, M or ZM

### PR DESCRIPTION
**Pull request by jgrocha**
## Previous discussion

The discussion started on the developers mailing list.

https://lists.osgeo.org/pipermail/qgis-developer/2024-July/066901.html

## Description

This PR changes the default WKT output to honor a space between the geometry type and the modifier Z, M, or ZM (when applied).

Example: `'PointZ (0 0 0)'` will be written as `'Point Z (0 0 0)'`.

Nothing was changed related to WKT parsing. Parsing is already quite flexible.
Different notations are allowed:
```python
print( QgsGeometry.fromWkt( 'Point (0 0 0)' ).asWkt() )
print( QgsGeometry.fromWkt( 'PointZ (0 0 0)' ).asWkt() )
print( QgsGeometry.fromWkt( 'Point Z(0 0 0)' ).asWkt() )
print( QgsGeometry.fromWkt( 'Point Z (0 0 0)' ).asWkt() )
```
Output based on this PR:
```
Point Z (0 0 0)
Point Z (0 0 0)
Point Z (0 0 0)
Point Z (0 0 0)
```
Output based on current master:
```
PointZ (0 0 0)
PointZ (0 0 0)
PointZ (0 0 0)
PointZ (0 0 0)
```

### Interoperability ensured by a new GDAL PR

The tools tested and discussed on the mailing list, namely, GDAL, GEOS, and QGIS support both forms for reading. 

The new https://github.com/OSGeo/gdal/pull/10370 enhances GDAL parsing to support also reading `PointZ` and `Point Z`.

With such enhancement, all three tools can interoperate based on WKT representations.

## Proposal

Since the interoperability is already ensured, we can decide the best way to write WKT representations.

The WKT specification is defined in the OGC document https://portal.ogc.org/files/?artifact_id=25355

According to the BNF Productions for Three-Dimension Geometry WKT (pages 55, 56), the WKT should be written as `point z <point z text>`, `linestring z <linestring z text>`, etc.

## Tests

Many tests use the WKT notation to compare generated and expected geometries. All tests were improved to support the new WKT writing.

### OS tools side by side

| QGIS            | GDAL            | shapely         | Postgis         | OpenLayers     | GEOS            |
|-----------------|-----------------|-----------------|-----------------|----------------|-----------------|
| `Point Z (1 2 3)` | `POINT Z (1 2 3)` | `POINT Z (1 2 3)` | `POINT Z (1 2 3)` | `POINT Z(1 2 3)` | `POINT Z (1 2 3)` |

QGIS

```python
a = QgsGeometry.fromWkt( 'Point Z (1 2 3)' )
print(a.asWkt())
```

GDAL

```python
from osgeo import ogr
a = ogr.CreateGeometryFromWkt("POINT Z (1 2 3)")
print(a.ExportToIsoWkt())
```

Shapely
```python
import shapely
a = shapely.from_wkt('POINT Z (1 2 3)')
print(shapely.to_wkt(a))
```

PostGIS
```SQL
select st_astext(st_pointfromtext('POINT Z (1 2 3)'));
```

OpenLayers

```javascript
import WKT from 'ol/format/WKT';
var format = new WKT();
let a = format.readFeature('POINT Z (1 2 3)');
console.log(format.writeFeature(a));
```

GEOS
```bash
echo 'POINT Z (1 2 3)' | geosop -a stdin -f wkt
```

### CODE REVIEW

**Comment by nyalldawson** [comment]
@mhugent do you have any insights as to why the current form was originally chosen?

**Comment by lbartoletti** [comment]
@jgrocha this is a very good idea! I made several tests, not yet released, some months ago. QGIS WKT is not readable by some other software. So +1. Let me just a couple of times to retrieve/finish my comparison/test and to review your PR. Thanks!

File: src/core/geometry/qgsabstractgeometry.cpp [hunk 1]
```diff
@@ -211,10 +211,15 @@ double QgsAbstractGeometry::area() const
 QString QgsAbstractGeometry::wktTypeStr() const
 {
   QString wkt = geometryType();
+  QString sufixe = "";
```
Comment (nyalldawson) [inline src/core/geometry/qgsabstractgeometry.cpp:214]:
```suggestion
  QString suffix;
```

**Comment by rouault** [comment]
@jgrocha There are some remaining tests to adapt. Cf CI failures

**Comment by jgrocha** [comment]
> > Done at 23rd Contributors Meeting in Bucharest :-)
> 
> @jgrocha Is there a parallel one to the one in Bratislava?



> > Done at 23rd Contributors Meeting in Bucharest :-)
> 
> @jgrocha Is there a parallel one to the one in Bratislava?

27th‐Contributor‐Meeting‐in‐Bratislava :-)

**Comment by jgrocha** [comment]
I'm missing something because tests are still failing on the CI. All tests are running fine on my side. 

```
jgr@dragon:~/dev/cpp/build-qgis$ ctest -V -R test_core_geometry_curvepolygon
UpdateCTestConfiguration  from :/home/jgr/dev/cpp/build-qgis/DartConfiguration.tcl
Parse Config file:/home/jgr/dev/cpp/build-qgis/DartConfiguration.tcl
 Add coverage exclude regular expressions.
 Add coverage exclude: /CMakeFiles/CMakeTmp/
UpdateCTestConfiguration  from :/home/jgr/dev/cpp/build-qgis/DartConfiguration.tcl
Parse Config file:/home/jgr/dev/cpp/build-qgis/DartConfiguration.tcl
Test project /home/jgr/dev/cpp/build-qgis
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added setup test logGitStatus required by fixture SOURCETREE
Added cleanup test checkGitStatus required by fixture SOURCETREE
Added 2 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start   1: logGitStatus

1: Test command: /usr/bin/sh "/home/jgr/dev/cpp/QGIS/scripts/chkgitstatus.sh" "log"
1: Test timeout computed to be: 1500
1/3 Test   #1: logGitStatus ......................   Passed    0.04 sec
test 233
    Start 233: test_core_geometry_curvepolygon

233: Test command: /home/jgr/dev/cpp/build-qgis/output/bin/test_core_geometry_curvepolygon "-maxwarnings" "10000"
233: Test timeout computed to be: 1500
233: ********* Start testing of TestQgsCurvePolygon *********
233: Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 11.3.0), ubuntu 22.04
233: PASS   : TestQgsCurvePolygon::initTestCase()
233: PASS   : TestQgsCurvePolygon::testConstructor()
233: PASS   : TestQgsCurvePolygon::testCopyConstructor()
233: PASS   : TestQgsCurvePolygon::testClear()
233: PASS   : TestQgsCurvePolygon::testClone()
233: PASS   : TestQgsCurvePolygon::testEquality()
233: PASS   : TestQgsCurvePolygon::testSetExteriorRing()
233: PASS   : TestQgsCurvePolygon::testAddInteriorRing()
233: PASS   : TestQgsCurvePolygon::testRemoveInteriorRing()
233: PASS   : TestQgsCurvePolygon::testMixedRingTypes()
233: PASS   : TestQgsCurvePolygon::test3dRings()
233: PASS   : TestQgsCurvePolygon::testAreaPerimeterWithInteriorRing()
233: PASS   : TestQgsCurvePolygon::testInsertVertex()
233: PASS   : TestQgsCurvePolygon::testMoveVertex()
233: PASS   : TestQgsCurvePolygon::testDeleteVertex()
233: PASS   : TestQgsCurvePolygon::testNextVertex()
233: PASS   : TestQgsCurvePolygon::testVertexAngle()
233: PASS   : TestQgsCurvePolygon::testDeleteVertexRemoveRing()
233: PASS   : TestQgsCurvePolygon::testHasCurvedSegments()
233: PASS   : TestQgsCurvePolygon::testClosestSegment()
233: PASS   : TestQgsCurvePolygon::testBoundary()
233: PASS   : TestQgsCurvePolygon::testBoundingBox()
233: PASS   : TestQgsCurvePolygon::testBoundingBox3D()
233: PASS   : TestQgsCurvePolygon::testBoundingBoxIntersects()
233: PASS   : TestQgsCurvePolygon::testRoundness()
233: PASS   : TestQgsCurvePolygon::testDropZValue()
233: PASS   : TestQgsCurvePolygon::testDropMValue()
233: PASS   : TestQgsCurvePolygon::testToPolygon()
233: PASS   : TestQgsCurvePolygon::testSurfaceToPolygon()
233: PASS   : TestQgsCurvePolygon::testWKB()
233: PASS   : TestQgsCurvePolygon::testWKT()
233: PASS   : TestQgsCurvePolygon::testExport()
233: PASS   : TestQgsCurvePolygon::testExportOfCompoundCurveRing()
233: PASS   : TestQgsCurvePolygon::testCast()
233: PASS   : TestQgsCurvePolygon::removeInteriorRings_github_issue_49578()
233: PASS   : TestQgsCurvePolygon::cleanupTestCase()
233: Totals: 36 passed, 0 failed, 0 skipped, 0 blacklisted, 13ms
233: ********* Finished testing of TestQgsCurvePolygon *********
2/3 Test #233: test_core_geometry_curvepolygon ...   Passed    0.22 sec
test 2
    Start   2: checkGitStatus

2: Test command: /usr/bin/sh "/home/jgr/dev/cpp/QGIS/scripts/chkgitstatus.sh" "check"
2: Test timeout computed to be: 1500
3/3 Test   #2: checkGitStatus ....................   Passed    0.02 sec

The following tests passed:
        logGitStatus
        test_core_geometry_curvepolygon
        checkGitStatus

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   0.29 sec
```

**Comment by jgrocha** [comment]
> @jgrocha this is a very good idea! I made several tests, not yet released, some months ago. QGIS WKT is not readable by some other software. So +1. Let me just a couple of times to retrieve/finish my comparison/test and to review your PR. Thanks!

Can you take a look at this @lbartoletti ?

**Comment by lbartoletti** [comment]
> > @jgrocha this is a very good idea! I made several tests, not yet released, some months ago. QGIS WKT is not readable by some other software. So +1. Let me just a couple of times to retrieve/finish my comparison/test and to review your PR. Thanks!
> 
> Can you take a look at this @lbartoletti ?

yes! Sorry, was busy. I'll take a look, for sure, on monday.

**Comment by lbartoletti** [comment]
OK for this change @jgrocha 
But you have to fix tests for Oracle, at least, this [file](https://github.com/qgis/QGIS/blob/e688fb104eb3906f69a5051a3b0ce61f480eac43/tests/src/python/test_provider_oracle.py) and maybe other providers.

**Comment by troopa81** [comment]
> ORACLE tests are not running on our CI. Maybe @troopa81 can help with the CI.

Well, everything seems to work fine now :+1:

**Review by lbartoletti — APPROVED** [review]

**Pull request merged**